go/libkb/naclwrap.go

```diff
@@ -700,20 +708,61 @@ func (k NaclDHKeyPair) EncryptToString(plaintext []byte, sender GenericKey) (str
 	return PacketArmoredEncode(info)
 }
 
-func (k NaclDHKeyPair) SecretSymmetricKey(reason EncryptionReason) ([]byte, error) {
-
+func (k NaclDHKeyPair) SecretSymmetricKey(reason EncryptionReason) (NaclSecretBoxKey, error) {
 	if !k.CanDecrypt() {
-		return nil, NoSecretKeyError{}
+		return NaclSecretBoxKey{}, NoSecretKeyError{}
 	}
-	if len(reason) < 8 {
-		return nil, KeyGenError{Msg: "reason must be at least 8 bytes"}
+
+	return deriveSymmetricKeyFromAsymmetric(*k.Private, reason)
+}
+
+// Derive a symmetric key using HMAC(k, reason).
+// Suitable for deriving from an asymmetric encryption key.
+// For deriving from a shared encryption key, this output is too close
+// to something that might be used as a public authenticator.
+func deriveSymmetricKeyFromAsymmetric(inKey NaclDHKeyPrivate, reason EncryptionReason) (NaclSecretBoxKey, error) {
+	var outKey = [32]byte{}
+	if len(reason) < encryptionReasonMinLength {
+		return outKey, KeyGenError{Msg: "reason must be at least 8 bytes"}
+	}
+
+	mac := hmac.New(sha256.New, inKey[:])
+	_, err := mac.Write(reason.Bytes())
+	if err != nil {
+		return outKey, err
 	}
+	out := mac.Sum(nil)
 
-	mac := hmac.New(sha256.New, k.Private[:])
-	mac.Write(reason.Bytes())
-	symmetricKey := mac.Sum(nil)
+	if copy(outKey[:], out) != len(outKey) {
+		return outKey, KeyGenError{Msg: "derived key of wrong size"}
+	}
+
+	return outKey, nil
+}
+
+// Derive a symmetric key using HKDF.
+func DeriveSymmetricKey(inKey NaclSecretBoxKey, reason EncryptionReason) (NaclSecretBoxKey, error) {
+	if len(reason) < encryptionReasonMinLength {
+		return NaclSecretBoxKey{}, KeyGenError{Msg: "reason must be at least 8 bytes"}
+	}
+
+	hash := sha256.New
+	info := []byte(reason)
+
+	// Use nil salt
+	stream := hkdf.New(hash, inKey[:], nil, info)
+
+	var outKey NaclSecretBoxKey
+	nRead, err := io.ReadFull(stream, outKey[:])
+	if err != nil {
+		return NaclSecretBoxKey{}, KeyGenError{Msg: err.Error()}
+	}
+	if nRead != NaclSecretBoxKeySize {
+		return NaclSecretBoxKey{}, KeyGenError{
+			Msg: fmt.Sprintf("read n bytes %v != %v", nRead, NaclSecretBoxKeySize)}
+	}
```

Comment:
These aren't real IO errors, right? They could only happen if we read too much of the stream? (Also I think `ReadFull` will always return an error above, if it's read less than the intended number of bytes.)

Maybe we should panic on them, instead of asking the caller to handle these?